config_converter/config_converter.py

```diff
@@ -0,0 +1,213 @@
+
+    def _plugin_convert(self, d: config_pb2.Directive, plugin: str) -> dict:
+        """Uses mapping rules passed to function to map all fields."""
+        cur = {}
+        outer_map = self.outer_map[d.name]
+        main_map = self.main_map[plugin]
+        special_params = self.special_map[plugin]
+        supported_dirs = self.supported_dirs[plugin]
+        plugin_type_map = {'tail': 'file'}
+        outer_map_copy = copy.copy(outer_map)
+        # ensure all outer level params needed are there in directive
+        for p in d.params:
+            if p.name in outer_map_copy:
+                if p.name == '@type':
+                    cur[outer_map[p.name]] = plugin_type_map[p.value]
+                else:
+                    cur[outer_map[p.name]] = p.value
+                outer_map_copy.pop(p.name)
+        if outer_map_copy != {}:
+            print('invalid configuration')
+            sys.exit()
+        specific = {}
+        for p in d.params:
+            if p.name not in outer_map:
+                if p.name in self.unsupported_fields:
+                    print(f'parameter {p.name} cant be converted')
+                elif p.name in special_params:
+                    self._map_special_fields(specific, p, plugin)
+                elif p.name not in main_map:
+                    print(f'parameter {p.name} is unknown')
+                else:
+                    specific[main_map[p.name]] = p.value
+        for nd in d.directives:
+            if nd.name not in supported_dirs:
+                print(f'directive {nd.name} is not supported')
+            else:
+                for np in nd.params:
+                    self._map_special_fields(specific, np, plugin)
+        cur[f'{cur["type"]}_{d.name}_config'] = specific
+        return cur
+
+    def _map_special_fields(self, specific: dict, p: config_pb2.Param,
+                            plugin: str) -> None:
+        """Cases on p with all special_map dict, calls a map function."""
+        if p.name in self.special_map[plugin]:
+            if plugin == 'in_tail':
+                self.parser_dir(specific, p)
+
+    def _create_mapping_info(self) -> None:
+        """Initializes all the mapping rules."""
+        # each sub dict contains fields that stay in the current level only
+        self.main_map = {
+            'in_tail': {
+                'exclude_path': 'exclude_path',
+                'path': 'path',
+                'path_key': 'path_field_name',
+                'pos_file': 'checkpoint_file',
+                'refresh_interval': 'refresh_interval',
+                'rotate_wait': 'rotate_wait'
+            }
+        }
+        # each sub list contains fields that need to be treated differently
+        self.special_map = {
+            'in_tail': [
+                'format', 'format_firstline', '@type',
+                'multiline_flush_interval', 'expression'
+            ] + [f'format{i}' for i in range(1, 21)]
```

Comment:
Seemed that there is a specific reason that format supports up to 20 in fluentd - can you add a link to the documentation so the future readers would understand?

Reply:
Yep, good idea!